[website]: Adds page on understanding grafana dashboards (#26617)
ethereum/go-ethereum: docs/monitoring/understanding-dashboards.md

![The Grafana dashboard](/images/docs/grafana/dashboard.png)

Each panel in the dashboard tracks a different metric that can be used to understand some aspect of how a Geth node is behaving. There are three main categories of panel in the default dashboard: System, Network and Blockchain. The individual panels are explained in the following sections.

## What do the panels show?

### System

Panels in the System category track the impact of Geth on the local machine, including memory and CPU usage.

#### CPU

![The CPU panel](/images/docs/grafana/cpu.png)

The CPU panel shows how much CPU is being used as a percentage of one processing core (i.e. 100% means complete usage of one processing core, 200% means complete usage of two processing cores). There are three processes plotted on the figure. The total CPU usage by the entire system is plotted as `system`; the percentage of time that the CPUs are idle waiting for disk i/o operations is plotted as `iowait`; the CPU usage by the Geth process is plotted as `geth`.

#### Memory

![The Memory panel](/images/docs/grafana/memory.png)

> Memory tracks the amount of RAM being used by Geth. Three metrics are plotted: the cache size, i.e. the total RAM reserved for Geth (default 1024 MB) is plotted as `held`; the amount of the cache actually being used by Geth is plotted as `used`; the memory allocations being made is plotted as `alloc`. 

Reviewer: `the memory allocations being made is plotted as alloc.` I have a nitpick here but it's ok to ignore. 

The alloc actually represents the number of bytes allocated by system per second. Maybe we can improve the sentence a bit?

Author: ```suggestion
Memory tracks the amount of RAM being used by Geth. Three metrics are plotted: the cache size, i.e. the total RAM reserved for Geth (default 1024 MB) is plotted as `held`; the amount of the cache actually being used by Geth is plotted as `used`; the number of bytes being allocated by the system per second is plotted as `alloc`. 
```

#### Disk

Disk tracks the rate that data is written to (plotted as `write`) or read from (plotted as `read`) the hard disk in units of MB/s.

![The Disk panel](/images/docs/grafana/disk.png)

### Network

Panels in the Network category track the data flow in and out of the local node.

#### Traffic

The Traffic panel shows the rate of data ingress and egress for all subprotocols, measured in units of kB/s.

![The Traffic panel](/images/docs/grafana/traffic.png)

#### Peers

> The Peers panel shows the number of individual peers the local node is connected to. The number of times the local node dials to find new peers and the number of times information is served from the local node are also tracked in this panel.

Reviewer: `The number of times the local node dials to find new peers` is incorrect, this metric represent the number of dials issued by geth per second.

`the number of times information is served from the local node` is incorrect, this metric represent the number of external connection received per second

Author: ```suggestion
The Peers panel shows the number of individual peers the local node is connected to. The number of dials issued by Geth per second and the number of external connections received per second are also tracked in this panel.
```

![The ETH ingress rate panel](/images/docs/grafana/eth-ingress-rate.png)

#### ETH egress data rate

Egress is the process of data leaving the local node and being transferred to its peers. This panel shows the rate that data specifically using the eth subprotocol is leaving the local node in units of kB/s (kilobytes per second). Make sure your dashboard includes the latest version of the eth subprotocol!

![The ETH egress rate panel](/images/docs/grafana/eth-egress-rate.png)

#### ETH ingress traffic

Ingress is the process of data arriving at the local node from its peers. This panel shows a moment-by-moment snapshot of the amount of data that is arriving at the local node, specifically using the eth subprotocol, in units of GB (gigabytes). Make sure your dashboard includes the latest version of the eth subprotocol!

![The ETH ingress traffic panel](/images/docs/grafana/eth-ingress-traffic.png)

#### ETH egress traffic

> Egress is the process of data leaving the local node and being transferred to its peers. This panel shows a moment-by-moment snapshot of the amount of data that is leaving the local node, specifically using the eth subprotocol, in units of GB (gigabytes). Make sure your dashboard includes the latest version of the eth subprotocol!

Reviewer: `This panel shows a moment-by-moment snapshot of the amount of data that is leaving the local node`

Perhaps

`This panel shows a moment-by-moment snapshot of the amount of data that has left the local node`?

Author: ```suggestion
Egress is the process of data leaving the local node and being transferred to its peers. This panel shows a moment-by-moment snapshot of the amount of data that has left the local node, specifically using the eth subprotocol, in units of GB (gigabytes). Make sure your dashboard includes the latest version of the eth subprotocol!
```

![The Chain head panel](/images/docs/grafana/chain-head.png)

#### Transaction pool

Geth has a capacity for pending transactions defined by `--txpool.globalslots` (default is 5160). The number of slots filled with transactions is tracked as `slots`. The transactions in the pool are divided into pending transactions and queued transactions. Pending transactions are ready to be processed and included in a block, whereas queued transactions are those whose transaction nonces are out of sequence. Queued transactions can become pending transactions if transactions with the missing nonces become available. In the dashboard pending transactions are labelled as `executable` and queued transactions are labelled `gapped`. The subset of those global transactions that originated from the local node are tracked as `local`.

![The tx pool panel](/images/docs/grafana/tx-pool.png)

#### Block processing

The block processing panel tracks the time taken to complete the various tasks involved in processing each block, measured in microseconds or nanoseconds. Specifically, this includes:

- **execution**: time taken to execute the transactions in the block

> - **validation**: time taken to compute a new state root and compare it to the one that arrived in the block

Reviewer: validation doesn't include the time spent on the state root computation(storage trie hashing + account trie hashing)

Author: Is this accurate?
```suggestion
- **validation**: time taken to validate that the information in a received block body matches what is described in the block header. 
```

- **commit**: time taken to write the new block to the chain data
- **account read**: time taken to access account information from the state trie

> - **account update**: time taken to update a leaf in the state trie

Reviewer: time taken to flush dirty account objects into the state trie(account trie)

Author: ```suggestion
- **account update**: time taken to incorporate dirty account objects into the state trie (account trie)
```

> - **account hash**: time taken to generate a hash of an account's data

Reviewer: time taken to re-compute the new root hash of the state trie(account trie)

Author: ```suggestion
- **account hash**: time taken to re-compute the new root hash of the state trie (account trie)
```

> - **account commit**: time taken to write new account data into the state trie

Reviewer: time taken to commit the state trie into database

Author: ```suggestion
- **account commit**: time taken to commit the changes of state trie (account trie) into database
```

- **storage read**: time taken to access smart contract storage data from the storage trie

> - **storage update**: time taken to change a piece of smart contract storage data in the storage trie

Reviewer: time taken to flush dirty storage slots into the storage tries

> - **storage update**: time taken to change a piece of smart contract storage data in the storage trie

Author: ```suggestion
- **storage update**: time taken to incorporate dirty storage slots into the storage tries
```

> - **storage hash**: time taken to generate a new hash for modified smart contract storage data

Reviewer: time take to re-compute the new root hash of storage tries

Author: ```suggestion
- **storage hash**: time take to re-compute the new root hash of storage tries
```

> - **storage commit**: time taken to write modified smart contract storage data to the storage trie. 

Reviewer: time take to commit the changes of storage tries into database

Author: ```suggestion
- **storage commit**: time take to commit the changes of storage tries into database
```

- **snapshot account read**: time taken to read account data from a snapshot

> - **snapshot storage read**: time taken to read storage data from a  snapshot

Reviewer: extra the blank

Author: ```suggestion
- **snapshot storage read**: time taken to read storage data from a snapshot
```

> - **snapshot commit**: time taken to write data to a snapshot

Reviewer: time take to flush the dirty state data as a new snapshot

Author: ```suggestion
- **snapshot commit**: time take to flush the dirty state data as a new snapshot
```

![The block processing panel](/images/docs/grafana/block-processing.png)

#### Transaction processing

> The transaction processing panel tracks the time taken to complete the various tasks involved in processing each block, measured as a mean rate of events per second:

Reviewer: `The transaction processing panel tracks the time taken to complete the various tasks involved in processing each block` it's incorrect.

`The transaction processing panel tracks the time taken to complete the various tasks involved in validating them received from the network`

Author: ```suggestion
The transaction processing panel tracks the time taken to complete the various tasks involved in validating the transactions received from the network, measured as a mean rate of events per second:
```

- **known**: rate of new transactions arriving at the node that are ignored because the local node already knows about them.

> - **valid**: rate that node marks known transactions as valid

Reviewer: rate that node marks new received transactions as valid

Author: ```suggestion
- **valid**: rate that node marks received transactions as valid
```

> - **invalid**: rate that node marks known transactions as invalid

Reviewer: rate that node marks received transactions as invalid

Author: ```suggestion
- **invalid**: rate that node marks received transactions as invalid
```

> - **underpriced**: rate that node marks transactions paying insufficient gas as invalid

Reviewer: rate that node marks transactions paying too low gas price as rejected

Author: ```suggestion
- **underpriced**: rate that node marks transactions paying too low gas price as rejected
```

- **executable discard**: rate that valid transactions are dropped from the transaction pool, e.g. because it is already known.
- **executable replace**: rate that valid transactions are replaced with a new one from same sender with same nonce but higher gas
- **executable ratelimit**: rate that valid transactions are dropped due to rate-limiting
- **executable nofunds**: rate that valid transations are dropped due to running out of ETH to pay gas
- **gapped discard**: rate that queued transactions are discarded from the transaction pool
- **gapped replace**: rate that queued transactions are replaced with a new one from same sender with same nonce but higher gas
- **gapped ratelimit**: rate that queued transactions are dropped due to rate limiting
- **gapped nofunds**: rate that queued transactions are dropped due to running out of ETH to pay gas

![The tx processing panel](/images/docs/grafana/tx-processing.png)

#### Block propagation

> Block propagation metrics track the rate that the local node hears about, receives and broadcasts blocks. This includes:

Reviewer: We need to add a note that block propagation is totally disabled after The Merge. So all numbers are expected to be zero.

Author: ah yes, of course - thanks.
```suggestion

<Note>
Block propagation was disabled in Geth at The Merge. Block propagation is now the responsibility of the consensus client.  Included here for archival interest.
</Note>
 
Block propagation metrics track the rate that the local node hears about, receives and broadcasts blocks. This includes:
```

> - **ingress announcements**: counts the number of inbound announcements processed by the local node. Announcements are messages from peers that signal that they have a block to share.

Reviewer: the number of inbound announcements per second

Author: ```suggestion
- **ingress announcements**: the number of inbound announcements per second. Announcements are messages from peers that signal that they have a block to share.
```

> - **known announcements**: counts the announcements that are ignored because the local node is already aware of them.

Reviewer: The number of announcements per second

Author: ```suggestion
- **known announcements**: the number of announcements per second
```

> - **malicious announcements**: counts the number of announcements from peers that are determined to be malicious, e.g. because they are trying to mount a denial-of-service attack on the local node.

Reviewer: the number of announcements per second

> - **ingress broadcasts**: Blocks downloaded from peers

Reviewer: The number of blocks directly propagated to local node per second.

Author: ```suggestion
- **ingress broadcasts**: the number of blocks directly propagated to local node per second.
```

- **known broadcasts**: counts all blocks that have been broadcast by peers including those that are too far behind the head to be downloaded.

> - **malicious broadcasts**: counts all blocks that are determined to be malicious

Reviewer: The number of blocks which are determined to be malicious per second

Author: ```suggestion
- **malicious broadcasts**: the number of blocks which are determined to be malicious per second
```

#### Transaction propagation

Transaction propagation tracks the sending and receiving of transactions on the peer-to-peer network. This includes:

> - **ingress announcements**: counts inbound announcements (notifications of a transaction's availability)

Reviewer: per second

Author: ```suggestion
- **ingress announcements**: counts inbound announcements (notifications of a transaction's availability) per second
```

> - **known announcements**: counts announcements that are ignored because the local node is already aware of them

Reviewer: per second

Author: ```suggestion
- **known announcements**: counts announcements that are ignored because the local node is already aware of them, per second
```

> - **underpriced announcements**: counts all announcements that never get fetched because they pay too little gas 

Reviewer: per second, too low gas price

Author: ```suggestion
- **underpriced announcements**: counts announcements per second that do not get fetched because they pay too little gas 
```

> - **malicious announcements**: counts announcements that are dropped because they appear malicious 

Reviewer: per second

Author: ```suggestion
- **malicious announcements**: counts announcements per second that are dropped because they appear malicious 
```

> - **ingress broadcasts**: counts all transactions fetched from peers

Reviewer: the number of transactions propagated from peers per second

Author: ```suggestion
- **ingress broadcasts**: the number of transactions propagated from peers per second
```

- **known broadcasts**: counts transactions that are ignored because they duplicate transactions that the local node already knows about
- **underpriced broadcasts**: counts all fetched transactions that are dropped due to paying insufficient gas
- **otherreject broadcasts**: counts transactions that are rejected for reasons other than paying too little gas
- **finished requests**: counts the successful delivery of a transaction, meaning it has been added to the local transaction pool
- **failed requests**: counts the number of failed transaction deliveries, e.g. because a peer disconnected unexpectedly
- **timed out requests**: counts the number of transaction requests that time out
- **ingress replies**: tracks the total number of inbound replies to requests for transactions
- **known replies**: tracks number of replies that are dropped because they are already known to the local node
- **underpriced replies**: counts the number of replies that get dropped due to paying too little gas

> - **otherreject replies**: coutns the number of replies to transaction requests that get dropped for reasons other than paying too little gas

Reviewer: All of them are measure as speeds, the number of XX per second.

![The tx propagation panel](/images/docs/grafana/tx-propagation.png)

#### Block forwarding

The block forwarding panel counts the announcements and the blocks that the local node receives that it should pass on to its peers. 

#### Transaction fetcher peers

The transaction fetcher peers panel shows how many peers the local node is connected to that can serve requests for transactions. The adjacent transaction fetcher hashes panel shows how many transaction hashes are available for fetching. Three statuses are reported in each panel: Waiting, queuing and fetching.

#### Reorg

The `reorg meter` panel simply counts the blocks added and the blocks removed during chain reorgs. The adjacent `Reorg total` panel shows the total number of reorg executions including both additions and removals.

#### Goroutines

> Tracks the total number of active goroutines being used by Geth. Goroutines are lighweight threads managed by the Go runtime, they allow processes to 

Reviewer: Ah, we should move it to System category?

Author: done!

- **leveldb read**: Rate that data is read to the fast-access levelDB database that stores recent data.
- **leveldb write**: Rate that data is written to the fast-access levelDB database that stores recent data.
- **ancient read**: Rate that data is read from the freezer (the database storing older data).
- **ancient write**: Rate that data is written to the freezer (the database storing older data)
- **compaction read**: Rate that data is written to the levelDb database while it is being compacted (i.e. free space is reclaimed by deleting uneccessary data)
- **compaction write**: Rate that data is read from to the levelDB database while it is being compacted (i.e. free space is reclaimed by deleting uneccessary data)
  
#### Session totals

Instead of the *rate* that data is read from, and written to, the levelDB and ancients databases (as per `Data rate`), this panel tracks the total amount of data read and written across the entire time Geth is running.

#### Persistent size

This panel shows the amount of data, in GB, in the levelDB and ancients databases.

#### Compaction time, delay and count

> These panels show the amount of time spent compacting the levelDB database, duration write operations to the database are delayed due to compaction and the frequency of compaction executions.

Reviewer: the count of various types of compaction executions.

Author: ```suggestion
These panels show the amount of time spent compacting the levelDB database, duration write operations to the database are delayed due to compaction and the count of various types of compaction executions.
```
